crates/cli/src/main.rs

```diff
@@ -1,57 +1,242 @@
+        let mut num_fd = 2;
+        for socket in tcpsockets {
+            num_fd += 1;
+            wasi_bldr = wasi_bldr.preopened_socket(num_fd, socket)?;
+        }
+
+        for (dir_name, dir) in preopened_dirs {
+            wasi_bldr = wasi_bldr.preopened_dir(dir, dir_name)?;
+        }
+
+        let wasi = wasi_bldr.build();
+
+        let mut store = wasmi::Store::new(engine, wasi);
+        wasmi_wasi::define_wasi(&mut linker, &mut store, |ctx| ctx)?;
+
+        Ok((linker, store))
+    }
+
+    /// Loads the Wasm [`Func`] from the given `wasm_bytes` with `wasi` linked.
+    ///
+    /// Returns the [`Func`] together with its [`Store`] for further processing.
+    ///
+    /// # Errors
+    ///
+    /// - If the Wasm module fails to parse or validate.
+    /// - If there are errors linking `wasi`.
+    /// - If the Wasm module fails to instantiate or start.
+    /// - If the Wasm module does not have an exported function `func_name`.
+    fn load_wasm_func_with_wasi(&self) -> Result<(String, Func, Store<WasiCtx>)> {
+        let engine = wasmi::Engine::default();
+        let module = load_wasm_module(&self.wasm_file, &engine)?;
+        let (linker, mut store) = self.link_wasi(&engine)?;
+        let (name, func) = self.load_func(&linker, &mut store, &module)?;
+        Ok((name, func, store))
+    }
+
+    fn load_func<T>(
+        &self,
+        linker: &Linker<T>,
+        mut store: &mut Store<T>,
+        module: &Module,
+    ) -> Result<(String, Func)> {
+        let instance = linker
+            .instantiate(&mut store, module)
+            .and_then(|pre| pre.start(&mut store))
+            .map_err(|error| anyhow!("failed to instantiate and start the Wasm module: {error}"))?;
+
+        if let Some(name) = &self.invoke {
+            // if a func name is provided
+            let func = instance
+                .get_export(&store, name)
+                .and_then(|ext| ext.into_func())
+                .ok_or_else(|| {
+                    let exported_funcs = display_exported_funcs(module);
+                    anyhow!(
+                        "could not find function \"{name}\" in {}\n\n{exported_funcs}",
+                        self.wasm_file.display()
+                    )
+                })?;
+            Ok((name.into(), func))
+        } else {
+            let (name, ext) = {
+                if let Some(ext) = instance.get_export(&mut store, "") {
+                    // try " "
+                    ("", ext)
+                } else if let Some(ext) = instance.get_export(&mut store, "_start") {
+                    ("_start", ext)
+                } else {
+                    // no function invoked plus no default function exported: we bail out
+                    let exported_funcs = display_exported_funcs(module);
+                    bail!(
+                        "missing function name argument for {}\n\n{exported_funcs}",
+                        self.wasm_file.display()
+                    )
```

Comment:
This code is duplicated within the same function and could be pulled into a helper function or closure that does the job without duplicating the code twice.